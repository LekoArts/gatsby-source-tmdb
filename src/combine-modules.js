const defaultModules = {
  account: {
    activate: true,
    endpoints: {
      tvs: ['accountFavoriteTv', 'accountRatedTv', 'accountTvWatchlist'],
      movies: ['accountFavoriteMovies', 'accountRatedMovies', 'accountMovieWatchlist'],
      list: 'accountLists',
    },
  },
  misc: {
    activate: false,
    endpoints: [
      ['miscUpcomingMovies'],
      ['miscNowPlayingMovies'],
      ['miscPopularMovies', 2],
      ['miscTopRatedMovies', 2],
      ['miscTopRatedTvs', 1],
      ['miscPopularTvs', 1],
    ],
  },
  tv: {
    activate: false,
    endpoints: [['tvAiringToday'], ['tvOnTheAir', 2]],
  },
}

/**
 * @name combineModules
 * @description The user could only put "activate" into an entry and
 * will get all "endpoints".
 * His/her override of the "activate" will be combined with the default "endpoints"
 * @param userModules
 * @returns Object - Combined Object
 */

const combineModules = userModules =>
  Object.assign(defaultModules, {
    account: {
      ...defaultModules.account,
      ...userModules.account,
    },
    misc: {
      ...defaultModules.misc,
      ...userModules.misc,
    },
    tv: {
      ...defaultModules.tv,
      ...userModules.tv,
    },
  })

exports.combineModules = combineModules
