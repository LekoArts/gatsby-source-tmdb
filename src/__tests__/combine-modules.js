import { defaultModules, combineModules } from '../combine-modules'

describe('combine modules', () => {
  it('should use default modules without user input', () => {
    expect(combineModules({})).toEqual(defaultModules)
  })
  it('should respect change of default movies values', () => {
    expect(
      combineModules({
        account: {
          activate: true,
          endpoints: {
            tvs: ['accountFavoriteTv'],
          },
        },
      })
    ).toMatchSnapshot()
  })
  it('should make use of the activate key', () => {
    expect(
      combineModules({
        account: {
          activate: false,
        },
        misc: {
          activate: true,
        },
        tv: {
          activate: true,
        },
      })
    ).toMatchSnapshot()
  })
  it('should allow changing "misc"', () => {
    expect(
      combineModules({
        account: {
          activate: false,
        },
        misc: {
          activate: true,
          endpoints: [['miscUpcomingMovies'], ['miscPopularMovies', 5]],
        },
      })
    ).toMatchSnapshot()
  })
})
