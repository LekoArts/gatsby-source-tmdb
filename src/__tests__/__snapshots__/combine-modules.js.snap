// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`combine modules should allow changing "misc" 1`] = `
Object {
  "account": Object {
    "activate": false,
    "endpoints": Object {
      "list": "accountLists",
      "movies": Array [
        "accountFavoriteMovies",
        "accountRatedMovies",
        "accountMovieWatchlist",
      ],
      "tvs": Array [
        "accountFavoriteTv",
        "accountRatedTv",
        "accountTvWatchlist",
      ],
    },
  },
  "misc": Object {
    "activate": true,
    "endpoints": Array [
      Array [
        "miscUpcomingMovies",
      ],
      Array [
        "miscPopularMovies",
        5,
      ],
    ],
  },
  "tv": Object {
    "activate": false,
    "endpoints": Array [
      Array [
        "tvAiringToday",
      ],
      Array [
        "tvOnTheAir",
        2,
      ],
    ],
  },
}
`;

exports[`combine modules should make use of the activate key 1`] = `
Object {
  "account": Object {
    "activate": false,
    "endpoints": Object {
      "list": "accountLists",
      "movies": Array [
        "accountFavoriteMovies",
        "accountRatedMovies",
        "accountMovieWatchlist",
      ],
      "tvs": Array [
        "accountFavoriteTv",
        "accountRatedTv",
        "accountTvWatchlist",
      ],
    },
  },
  "misc": Object {
    "activate": true,
    "endpoints": Array [
      Array [
        "miscUpcomingMovies",
      ],
      Array [
        "miscNowPlayingMovies",
      ],
      Array [
        "miscPopularMovies",
        2,
      ],
      Array [
        "miscTopRatedMovies",
        2,
      ],
      Array [
        "miscTopRatedTvs",
        1,
      ],
      Array [
        "miscPopularTvs",
        1,
      ],
    ],
  },
  "tv": Object {
    "activate": true,
    "endpoints": Array [
      Array [
        "tvAiringToday",
      ],
      Array [
        "tvOnTheAir",
        2,
      ],
    ],
  },
}
`;

exports[`combine modules should respect change of default movies values 1`] = `
Object {
  "account": Object {
    "activate": true,
    "endpoints": Object {
      "tvs": Array [
        "accountFavoriteTv",
      ],
    },
  },
  "misc": Object {
    "activate": false,
    "endpoints": Array [
      Array [
        "miscUpcomingMovies",
      ],
      Array [
        "miscNowPlayingMovies",
      ],
      Array [
        "miscPopularMovies",
        2,
      ],
      Array [
        "miscTopRatedMovies",
        2,
      ],
      Array [
        "miscTopRatedTvs",
        1,
      ],
      Array [
        "miscPopularTvs",
        1,
      ],
    ],
  },
  "tv": Object {
    "activate": false,
    "endpoints": Array [
      Array [
        "tvAiringToday",
      ],
      Array [
        "tvOnTheAir",
        2,
      ],
    ],
  },
}
`;
